### ISSUE TITLE
Confused on how to create findOrCreate syntax with OR condition, promise style

### ISSUE THREAD

**Issue by tommedema** [sequelize/sequelize#3805]
I'm trying to create the following function: if an `Url` exists with either its `entryUrl` or `exitUrl` being equal to `strUrl`, then return the matched `Url` instance. Otheriwse, create a new instance and return the created instance.

From my understanding of promises the following should work:

	/**
	 * Saves a new url associated with a given post, if the url is unique. Otherwise, resolves to the found url.
	 * @param {Object} parsedUrl - a parsed node url object.
	 * @returns {Promise} promise - a promise that resolves when the post url has been saved or found.
	 */
	exports.findOrCreatePostUrl = function(parsedUrl, post)
	{
		var strUrl = parsedUrl.format();

		return Url.findOne({
			where: {
				exitUrl: strUrl
			}
		})
		.then(function(url)
		{
			if (!url)
			{
				return Url.findOrCreate({
					where: {
						entryUrl: strUrl
					},
					defaults: {
						entryUrl: strUrl
					}
				})
				.then(function(url)
				{
					return post.addUrl(url);
				});
			}

                        //what to return here?
		});

	};

The idea being that if `Url.findOne` resolves to no result, we create a new url as long as there isn't one already with a matching `entryUrl`.

However when I use it like so:

	api.findOrCreatePostUrl(parsedUrl, post)
	.then(function(url)
	{
		log.info('the url is', url);
	})
	.catch(function(err)
	{
		log.error('catched error', err);
	});

The result is a bit unexpected. No matter whether a `Url` is created or not, it always resolved and the output is always: `the url is [ [ 1 ] ]`, rather than an actual object. Also, I'm not sure what to return at the location of the comment above.

In short, I'm wondering how a findOrCreate method with an OR condition should be implemented.

**Comment by BridgeAR** [comment]
In the comment you want to write
```js
return url;
```
And the return you see is from:
```js
return post.addUrl(url);
```
I'm not sure what post is, so I can't say for sure what you want.

**Comment by mickhansen** [comment]
You need to use `spread` with `findOrCreate` since the result is `[instances, created]`. I'm not a 100% sure what the return value is for `addAssociation` in that case.

As for findOrCreate with $or you should be able to do:

```js
Url.findOrCreate({
  where: {
    $or: {
      entryUrl: url,
      exitUrl: url
    }
  },
  defaults: {
    entryUrl: url
  }
});
```

**Comment by mickhansen** [comment]
Might not work with `$or` actually since sequelize checks for plain objecst atm, but it should work with `where: sequelize.or()`

**Comment by tommedema** [comment]
I tried both $or and sequelize.or, both didn't work for me.

**Comment by mickhansen** [comment]
https://github.com/sequelize/sequelize/blob/master/lib/model.js#L1187 
`sequelize.or()` shoul definitely work.

**Comment by tommedema** [comment]
@BridgeAR  can I make it resolve to the created `url` rather than the return from `post.addUrl(url)` ?

**Comment by mickhansen** [comment]
@tommedema You'll have to be more precise to "didnt work". Show SQL logs etc.

**Comment by tommedema** [comment]
@mickhansen you're right that it works like this, I must have mistyped. This works:

	exports.findOrCreatePostUrl = function(parsedUrl, post)
	{
		var strUrl = parsedUrl.format();

		return Url.findOrCreate({
			where: Sequelize.or(
				{
					entryUrl: strUrl
				},
				{
					exitUrl: strUrl
				}
			),
			defaults: {
				entryUrl: strUrl
			}
		})
		.then(function(url)
		{
			return post.addUrl(url);
		});

	};

However, is it possible to make the promise resolve to the created  or found `url`, rather than the result from `post.addUrl(url)`?

**Comment by mickhansen** [comment]
Use `tap` instead of `then` or use `.return(url)` after `addUrl`.
(more of a promise thing than sequelize)
